### DISCUSSION TITLE
Singular Matrix Error in Flow past a cylinder problem

### DISCUSSION THREAD

**Discussion by ChrocheMisawa** [idaholab/moose#17760] [Q&A Modules: Navier-Stokes] [answered]
Hello,
 
      I'm setting up a similar simulation to #17470 except that I am doing a 2D problem for now.
      The geometry is set up as:
![image](https://user-images.githubusercontent.com/47645670/117350099-27afae00-ae7a-11eb-9eb9-4d0bd5e98b9d.png)
 
      Also, uniform refinement will be applied in MOOSE input for convergence purpose. All my input files used are attached in the zip file.

      First, I started with a FEM approach (Flow_Past_a_Cylinder_FE_ins.i) which the results looks reasonable.
    
     Since my alternate goal is to model compressible flow, I then tried to run simulations with FVM approach for both incomprehensible and compressible flow.

     I used the Mr. Lindsay's (@lindsayad) branch for the compressible flow, the simulation proceed for few timestep but then ran into bad convergence. He did mention the transient feature is not fully supported yet but I'm not sure if the problem is more related to my problem setup.

     Then I used Mr. Giudicelli (@GiudGiud) branch for the incompressible flow simulation. I would like to do transient problem but I got this "Matrix A is singular!" error even for the steady state case:
![image](https://user-images.githubusercontent.com/47645670/117351400-af49ec80-ae7b-11eb-8fe9-395fdb1af7f7.png)

     The input file is adapted from modules\navier_stokes\test\tests\finite_volume\pins\channel-flow\heated\2d-rc-heated-boussinesq.i

     I wonder what could be wrong with my setup?

 
[Flow_Past_a_Cylinder.zip](https://github.com/idaholab/moose/files/6436968/Flow_Past_a_Cylinder.zip)

Thank you!
Zeyu Chen

**Comment by GiudGiud** [comment, marked as answer]
Hi Zeyu

I took a look at the inputs (the incompressible fv one) you had posted on the previous thread and I had a few comments:
- you should use `rc` interpolation for velocity instead of average.This will remove oscillations in pressure and improve convergence
- you should use the Newton solve type since INSFV uses automatic differentiation
- you should remove the forcing function kernels. These are used for MMS testing, not for real simulations
After that your simulation converged on my laptop

Best,
Guillaume

**Reply by ChrocheMisawa** [reply]
Hi Guillaume,

     I assume you are referring to the "Flow_Past_a_Cylinder_Fv_ins.i"

     I tried to modify the input file with your suggestions, after that,

     at first it says: 
        A 'NSFVMassFluxAdvection' is not a registered object.
        -- I guess this is because I switched to your branch, and I have to change it back to INSFVMassFluxAdvection
            (btw I can't pull #17538 anymore, do you have any idea why?)

     Now, the Singular Matrix Error is solved, but I'm still getting bad convergence even at the first timestep.




--- Zeyu Chen

**Reply by ChrocheMisawa** [reply]
---update:
  I used
     petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -sub_pc_factor_levels'
     petsc_options_value = 'asm      2               ilu          4'

  and now it seemed to converge for the first several "real time" seconds, then failed afterwards.

  By looking at the output, the simulation looks actually pretty good before reaching a "steady state".

  Do you have any idea why this is happening? I suspect it is related to the temperature.

 
[Flow_Past_a_Cylinder_fv_ins.zip](https://github.com/idaholab/moose/files/6437269/Flow_Past_a_Cylinder_fv_ins.zip)


Thank you,
Zeyu Chen

**Reply by GiudGiud** [reply]
It worked fine for the whole transient with the previous petsc settings. 
I was suspecting the pressure oscillations. The simulation seems converged to me, it just can't meet the convergence criterion for various reasons. 
Either way maybe just relax your tolerance with
``
nl_abs_tol = 1e-10
nl_rel_tol = 1e-10
``
or ``automatic_scaling = true`` to your executioner block, see if that helps

**Reply by ChrocheMisawa** [reply]
Yes, you are right looks like to relaxed my tolerance that made it run for 70~80 time steps.

I relaxed my tolerance to nl_abs_tol = 1e-8 nl_rel_tol = 1e-8 and it converges now.

Thank you!
Zeyu Chen
